File: packages/backend-python/toon_backend/routes/actions.py
"""
Action routes for agent endpoints
"""
import json
import time
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from typing import Dict, Any
from toon_backend.event_emitter import EventEmitter, EventType, AgentEvent


def create_action_handler(
    action: Dict[str, Any],
    event_emitter: EventEmitter,
    base_url: str = None
):
    """Create a FastAPI route handler for an action"""
    
    async def handler(request: Request):
        action_id = f"{action['id']}_{int(time.time() * 1000)}"
        start_time = time.time()
        
        # Emit action start event
        event_emitter.emit(AgentEvent(
            event_type=EventType.ACTION_START,
            event_id=action_id,
            timestamp=int(time.time() * 1000),
            actionId=action_id,
            actionName=action.get("name", ""),
            endpoint=action.get("endpoint", ""),
            method=action.get("method", "GET"),
            parameters={
                **dict(request.query_params),
                **dict(request.path_params),
            }
        ))
        
        try:
            # Build full URL if base_url is provided
            full_url = f"{base_url}{action['endpoint']}" if base_url else action['endpoint']
            
            # In a real implementation, this would make an actual HTTP request
            # For now, we'll simulate it
            response_data = await execute_action(action, request, full_url)
            
            duration = int((time.time() - start_time) * 1000)
            
            # Emit action end event
            event_emitter.emit(AgentEvent(
                event_type=EventType.ACTION_END,
                event_id=f"{action_id}_end",
                timestamp=int(time.time() * 1000),
                actionId=action_id,
                response=response_data,
                statusCode=200,
                success=True,
                duration=duration
            ))
            
            return response_data
        except Exception as error:
            duration = int((time.time() - start_time) * 1000)
            error_message = str(error)
            
            # Emit error event
            event_emitter.emit(AgentEvent(
                event_type=EventType.ACTION_END,
                event_id=f"{action_id}_end",
                timestamp=int(time.time() * 1000),
                actionId=action_id,
                success=False,
                error=error_message,
                duration=duration
            ))
            
            from fastapi.responses import JSONResponse
            return JSONResponse(
                content={"error": error_message},
                status_code=500
            )
    
    return handler


async def execute_action(
    action: Dict[str, Any],
    request: Request,
    full_url: str
) -> Dict[str, Any]:
    """Execute an action (mock implementation)"""
    # In a real implementation, this would:
    # 1. Build the request URL with query parameters
    # 2. Make an HTTP request to the actual backend
    # 3. Return the response
    
    return {
        "action": action.get("id", ""),
        "endpoint": action.get("endpoint", ""),
        "method": action.get("method", "GET"),
        "parameters": {
            **dict(request.query_params),
            **dict(request.path_params),
        },
        "message": "This is a mock response. In production, this would call the actual backend."
    }


def register_action_routes(
    router: APIRouter,
    schema: Dict[str, Any],
    event_emitter: EventEmitter
):
    """Register action routes from schema"""
    actions = schema.get("actions", [])
    base_url = schema.get("baseUrl", "")
    
    for action in actions:
        method = action.get("method", "GET").lower()
        endpoint = action.get("endpoint", "")
        handler = create_action_handler(action, event_emitter, base_url)
        
        # Register route based on method
        if method == "get":
            router.get(endpoint)(handler)
        elif method == "post":
            router.post(endpoint)(handler)
        elif method == "put":
            router.put(endpoint)(handler)
        elif method == "delete":
            router.delete(endpoint)(handler)
        elif method == "patch":
            router.patch(endpoint)(handler)
        elif method == "head":
            router.head(endpoint)(handler)
        elif method == "options":
            router.options(endpoint)(handler)

File: packages/backend-node/src/routes/actions.ts
import { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import { AgentSchema, AgentAction, EventType } from '@toon/core';
import { EventEmitter } from '../event-emitter.js';

interface ActionRequest {
  Params: {
    actionId: string;
  };
  Body?: unknown;
  Querystring?: Record<string, unknown>;
}

/**
 * Register action routes from agent schema
 */
export function registerActionRoutes(
  fastify: FastifyInstance,
  schema: AgentSchema,
  eventEmitter: EventEmitter
): void {
  for (const action of schema.actions) {
    const routeHandler = createActionHandler(action, eventEmitter, schema.baseUrl);

    // Register route based on method
    switch (action.method) {
      case 'GET':
        fastify.get(action.endpoint, routeHandler);
        break;
      case 'POST':
        fastify.post(action.endpoint, routeHandler);
        break;
      case 'PUT':
        fastify.put(action.endpoint, routeHandler);
        break;
      case 'DELETE':
        fastify.delete(action.endpoint, routeHandler);
        break;
      case 'PATCH':
        fastify.patch(action.endpoint, routeHandler);
        break;
      case 'HEAD':
        fastify.head(action.endpoint, routeHandler);
        break;
      case 'OPTIONS':
        fastify.options(action.endpoint, routeHandler);
        break;
    }
  }
}

function createActionHandler(
  action: AgentAction,
  eventEmitter: EventEmitter,
  baseUrl?: string
) {
  return async (request: FastifyRequest<ActionRequest>, reply: FastifyReply) => {
    const actionId = `${action.id}_${Date.now()}`;
    const startTime = Date.now();

    // Emit action start event
    eventEmitter.emit({
      type: EventType.ACTION_START,
      timestamp: Date.now(),
      id: actionId,
      actionId,
      actionName: action.name,
      endpoint: action.endpoint,
      method: action.method,
      parameters: {
        ...(request.query as Record<string, unknown>),
        ...(request.params as Record<string, unknown>),
        ...(request.body as Record<string, unknown>),
      },
    });

    try {
      // Build full URL if baseUrl is provided
      const fullUrl = baseUrl ? `${baseUrl}${action.endpoint}` : action.endpoint;

      // In a real implementation, this would make an actual HTTP request
      // For now, we'll simulate it
      const response = await executeAction(action, request, fullUrl);

      const duration = Date.now() - startTime;

      // Emit action end event
      eventEmitter.emit({
        type: EventType.ACTION_END,
        timestamp: Date.now(),
        id: `${actionId}_end`,
        actionId,
        response,
        statusCode: 200,
        success: true,
        duration,
      });

      return reply.send(response);
    } catch (error) {
      const duration = Date.now() - startTime;
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';

      // Emit error event
      eventEmitter.emit({
        type: EventType.ACTION_END,
        timestamp: Date.now(),
        id: `${actionId}_end`,
        actionId,
        success: false,
        error: errorMessage,
        duration,
      });

      return reply.code(500).send({ error: errorMessage });
    }
  };
}

async function executeAction(
  action: AgentAction,
  request: FastifyRequest<ActionRequest>,
  fullUrl: string
): Promise<unknown> {
  // In a real implementation, this would:
  // 1. Build the request URL with query parameters
  // 2. Make an HTTP request to the actual backend
  // 3. Return the response

  // For now, return a mock response
  return {
    action: action.id,
    endpoint: action.endpoint,
    method: action.method,
    parameters: {
      ...(request.query as Record<string, unknown>),
      ...(request.params as Record<string, unknown>),
      ...(request.body as Record<string, unknown>),
    },
    message: 'This is a mock response. In production, this would call the actual backend.',
  };
}

